cudaaligner/src/alignment_impl.cpp

```diff
@@ -41,36 +41,39 @@ char alignment_state_to_cigar_state(AlignmentState s)
     case AlignmentState::mismatch: return 'M';
     case AlignmentState::insertion: return 'I';
     case AlignmentState::deletion: return 'D';
-    default: throw std::runtime_error("Unrecognized alignment state.");
+    default: assert(false); return '!';
     }
 }
 
-} // namespace
-
-AlignmentImpl::AlignmentImpl(const char* query, int32_t query_length, const char* target, int32_t target_length)
-    : query_(query, query + throw_on_negative(query_length, "query_length has to be non-negative."))
-    , target_(target, target + throw_on_negative(target_length, "target_length has to be non-negative."))
-    , status_(StatusType::uninitialized)
-    , type_(AlignmentType::unset)
-    , alignment_()
-    , is_optimal_(false)
+char alignment_state_to_cigar_state_extended(AlignmentState s)
 {
-    // Initialize Alignment object.
+    // CIGAR string format from http://bioinformatics.cvr.ac.uk/blog/tag/cigar-string/
+    // Implementing a reduced set of CIGAR states, covering only the M, D and I characters.
+    switch (s)
+    {
+    case AlignmentState::match: return '=';
+    case AlignmentState::mismatch: return 'X';
+    case AlignmentState::insertion: return 'I';
+    case AlignmentState::deletion: return 'D';
+    default: assert(false); return '!';
+    }
 }
 
-std::string AlignmentImpl::convert_to_cigar() const
+template <typename CigarConversionFunction>
+std::string convert_to_cigar_impl(const std::vector<AlignmentState>& alignment, CigarConversionFunction convert_to_cigar)
```

Comment:
Is there a reason for not making this a private method and using `alignment_`?

Reply:
I don't see a benefit of making this part of the class. I view it as an implementation, which is not really part part of the class.
As a separate helper function, it is clear what data it operates on and that this data remains constant. Since it is in an anonymous namespace, it is not exposed in any way - but really just the implementation of the public `convert_to_cigar()` member function.
If it was a class member itself, the helper function could access and alter all data members.

Reply:
Ok, I agree